src/miniscript/evaluate.rs

```diff
@@ -0,0 +1,606 @@
+//! # Interpreatation of Miniscript scripts
+//!
+//! Traits and implementations to support for interpretation of miniscripts.
+//!
+//!
+use bitcoin_hashes::sha256;
+use secp256k1;
+
+use Error;
+use miniscript::astelem::AstElem;
+use ToPublicKey;
+use secp256k1::{Secp256k1, Signature, VerifyOnly};
+use bitcoin_hashes::Hash;
+use bitcoin;
+
+///Primitives for Miniscript: a vector of public keys(sigs), hashlocks and timelocks
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct Primitives{
+    pub pks: Vec<bitcoin::PublicKey>,
+    pub hashlocks : Vec<Vec<u8>>,
+    pub timelocks : Vec<u32>
+}
+
+impl Primitives
+{
+    fn join(&mut self, other: &Primitives){
+        self.pks.extend(other.pks.clone());
+        self.hashlocks.extend(other.hashlocks.clone());
+        self.timelocks.extend(other.timelocks.clone());
+    }
+}
+
+
+/// Definition of Stack Element of the Witness stack used for interpretation of Miniscript.
+/// All stack elements with vec![] go to Dissatisfied and vec![1] are marked to Satisfied.
+/// Others are directly pushed as witness
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub enum StackElement {
+    /// Result of a satisfied E/T/W/F
+    Satisfied,
+    /// Result of a dissatisfied E/W
+    Dissatisfied,
+    /// Input from the witness stack
+    Witness(Vec<u8>),
+}
+
+impl From<Vec<u8>> for StackElement {
+    fn from(v: Vec<u8>) -> StackElement {
+        match v{
+            _ if v == vec![1] => StackElement::Satisfied,
+            _ if v == vec![] => StackElement::Dissatisfied,
+            _ => StackElement::Witness(v)
+        }
+    }
+}
+
+impl StackElement{
+
+    pub fn is_witness_or_err(self) -> Result<Vec<u8>, Error>
```

Comment:
I think you should implement this on `Option<StackElement>`, have it take `&self` rather than `self`, and return a `&[u8]` rather than `Vec<u8>`.

Reply:
and rename the method to just `witness`. The `_or_err` is implied by the type.

Reply:
Similarly you should have a method `preimage(hash: sha256d::Hash)` which returns a preimage if it satisfies the length/hash check